benzina/src/json/convert.rs

```diff
@@ -0,0 +1,123 @@
+///     last_name: String,
+///     #[diesel(deserialize_as = Jsonb<UserPermissions>)]
+///     permissions: UserPermissions,
+///     #[diesel(deserialize_as = NullableJsonb<UserPreferences>)]
+///     preferences: Option<UserPreferences>,
+/// }
+///
+/// #[derive(Debug, Insertable)]
+/// #[diesel(table_name = users)]
+/// struct NewUser {
+///     first_name: String,
+///     last_name: String,
+///     #[diesel(serialize_as = Jsonb<UserPermissions>)]
+///     permissions: UserPermissions,
+///     #[diesel(serialize_as = NullableJsonb<UserPreferences>)]
+///     preferences: Option<UserPreferences>,
+/// }
+///
+///
+/// #[derive(Debug)]
+/// struct UserPermissions {
+///     can_delete: bool,
+///     can_update: bool,
+///     can_read: bool,
+/// }
+///
+/// #[derive(Debug)]
+/// struct UserPreferences {
+///     theme_color: [u8; 3],
+///     personal_kv: BTreeMap<String, String>,
+/// }
+///
+/// diesel::table! {
+///     users (id) {
+///         id -> Int4,
+///         first_name -> Text,
+///         last_name -> Text,
+///         permissions -> Jsonb,
+///         preferences -> Nullable<Jsonb>,
+///     }
+/// }
+///
+///
+/// json_wrapped!(UserPermissions);
+/// ```
+///
+/// `Json` works in the same exact way of `Jsonb`.
+#[macro_export]
+macro_rules! json_wrapped {
+    (
+        $($type:ty),*
+    ) => {
+        $(
+            impl $crate::__private::std::convert::From<$crate::Jsonb<$type>> for $type {
+                fn from(value: $crate::Jsonb<$type>) -> Self {
+                    $crate::Jsonb::into_inner(value)
+                }
+            }
+
+            //impl $crate::__private::std::convert::From<$crate::__private::std::option::Option<$type>> for $crate::NullableJsonb<$type> {
+            //    fn from (value: $crate::__private::std::option::Option<$type>) -> Self {
+            //        $crate::NullableJsonb::new(value)
+            //    }
+            //}
+
+            impl $crate::__private::std::convert::From<$crate::Json<$type>> for $type {
+                fn from(value: $crate::Json<$type>) -> Self {
+                    $crate::Json::into_inner(value)
+                }
+            }
+
+            //impl $crate::__private::std::convert::From<$crate::__private::std::option::Option<$type>> for $crate::NullableJson<$type> {
+            //    fn from (value: $crate::__private::std::option::Option<$type>) -> Self {
+            //        $crate::NullableJson::new(value)
+            //    }
+            //}
+        )*
+    };
+}
+
+pub(super) fn sql_serialize<T>(
+    value: &T,
+    out: &mut diesel::serialize::Output<'_, '_, Pg>,
+) -> diesel::serialize::Result
+where
+    T: Serialize,
+{
+    serde_json::to_writer(out, value)
+        .map(|()| IsNull::No)
+        .map_err(Into::into)
+}
+
+pub(super) fn sql_deserialize<T>(value: PgValue<'_>) -> diesel::deserialize::Result<T>
+where
+    T: DeserializeOwned,
+{
+    let inner = serde_json::from_slice(value.as_bytes())?;
+
+    Ok(inner)
```

Comment:
Uses the same compact format as `sql_serialize`.

```suggestion
    serde_json::from_slice(value.as_bytes()).map_err(Into::into)
```